Color mode

Anders-Holst opened the pull request:
As previously promised, adding color-mode functionality to the low-level ControlInterface, and the corresponding unit tests.

Anders-Holst committed:
Added color-mode functionality to ControlInterface

Anders-Holst committed:
Added unit tests for color-mode functionality

scrool commented on tests/test_control_low.py:
```diff
@@ -207,6 +207,67 @@ def test_mqtt(self):
+                },
+            )
+
+        res = ctr.set_led_color_hsv(180, 255, 255)._data
+        if self.isrecording:
+            print("set_led_color_hsv:", res)
+        else:
+            self.assertEqual(res, {"code": 1000})
```
Let's move these up where RGB are set.

Then in both `get_led_color()` and `get_mode()` we could test for these HSV values.

Anders-Holst replied:
But it does set the values before testing, with both tests! 
Note that there is only one color set at any one time on the leds, but this color can be specified *either* through RGB or HSV. So when setting the RGB to (128, 128, 0), i.e. dim yellow, the HSV coordinates are updated as well, to (60, 255, 128).  (Hue 60 means yellow, value 128 means half-strength). These values are then tested in the test. Then the HSV coordinates are set to (180, 255, 255), ie full strength cyan (hue 180 means cyan, value 255 is full strength), which results in the RGB coordinates being correctly updated to (0, 255, 255). Which is again tested, but this time through `get_mode`.

scrool replied:
> Note that there is only one color set at any one time on the leds, but this color can be specified either through RGB or HSV.

Right, this is what I haven't realized. My bad. I'll explain this commented against line that failed.

scrool commented on tests/test_control_low.py:
```diff
@@ -207,6 +207,67 @@ def test_mqtt(self):
+                        "hue": 180,
+                        "saturation": 255,
+                        "value": 255,
+                        "red": 0,
+                        "green": 255,
+                        "blue": 255,
```
These values don't correspond to `set_led_color_rgb()` - was this recoded in single run?

Anders-Holst replied:
See reply above.

scrool reviewed (changes requested):
Tests without any changes pass on my machine as well. :+1: 

When I remove a casette:
```
rm -f tests/cassettes/TestControlInterface.test_color.yaml
```

change IP address to my device, set `isRecording = True` and run only one test (directly with `pytest`, not with `tox` this time):

```
$ pytest -v tests/test_control_low.py -k test_color
```

tests are correctly recorded. :+1: 

I _think_ rearanging of HSV tests to set them first before we test for values it could work even without isRecording.

scrool commented on tests/test_control_low.py:
```diff
@@ -207,6 +207,67 @@ def test_mqtt(self):
+                res,
+                {
+                    "hue": 60,
+                    "saturation": 255,
```
For some reason on the device I tested this ([Model TWS250STP-BEU](https://xled-docs.readthedocs.io/en/latest/hardware.html#model-tws250stp-beu)) with FW version 2.8.3 this test fails - expects `255` but device returns `254`.

Anders-Holst replied:
@scrool, this is strange. I have the exact same model (250 leds, profile RGB), but with version 2.7.1. It is the one used to produce the cassettes. 
So what you say is that with firmware 2.8.3, when setting RGB to (128, 128, 0), the HSV becomes (60, 254, 128) ?
Whereas with version 2.7.1 the same input results in HSV (60, 255, 128). That's strange, since 255 should be the correct number: if any RGB component is zero (but not all), then saturation is at maximum, which is 255. Maybe they adjusted the calculation and accidentally introduced a rounding error?  
So I will upgrade my leds too and see if it changes here too. We know of course that different models and versions may differ in their output. But if this is a rounding error in the device, then I can try to find another test that is not "on the border" but gives the same value in both these versions.
(And then it suddenly  dawns on me, that RGBW and AWW models surely are different in these parameters. That is, the current documentation are limited to profile RGB right now, until someone with access to such devices can check what comes out.)

scrool replied:
That is strange indeed. I have downgraded FW on that device to 2.7.1 and that test still fails the same way.

Btw. I have also some RGBW here, I just didn't want to screw up my config :)

btw. using [shell wrappers](https://github.com/xled/xled-shell) it looks like this:

```
$ echo '{"red": 128, "green": 128, "blue": 0}' | ./post-json $IPADDR led/color && ./get $IPADDR led/color    
POST led/color
{
  "code": 1000
}
GET led/color
{
  "hue": 60,
  "saturation": 254,
  "value": 128,
  "red": 128,
  "green": 128,
  "blue": 0,
  "code": 1000
}
```

and to test that other way around:

```
$ echo '{"hue": 0, "saturation": 0, "value": 0}' | ./post-json $IPADDR led/color && ./get $IPADDR led/color
POST led/color
{
  "code": 1000
}
GET led/color
{
  "hue": 0,
  "saturation": 0,
  "value": 0,
  "red": 0,
  "green": 0,
  "blue": 0,
  "code": 1000
}
$ echo '{"hue": 60, "saturation": 255, "value": 128}' | ./post-json $IPADDR led/color && ./get $IPADDR led/color
POST led/color
{
  "code": 1000
}
GET led/color
{
  "hue": 60,
  "saturation": 255,
  "value": 128,
  "red": 128,
  "green": 128,
  "blue": 0,
  "code": 1000
}
```

Anders-Holst replied:
Yes, odd that it remains when you downgraded... it is contrary to what I get: 
I have now upgraded one of my two lights, both of exactly the same model TWS250STP-BEU. Then I tested on both, and the new firmware really has some problems with color. Check these identical sequences on my two devices:
```
>>> ctr1.firmware_version()._data
{'version': '2.7.1', 'code': 1000}
>>> ctr1.set_led_color_rgb(0, 255, 255)._data
{'code': 1000}
>>> ctr1.get_led_color()._data
{'hue': 180, 'saturation': 255, 'value': 255, 'red': 0, 'green': 255, 'blue': 255, 'code': 1000}
>>> ctr1.set_led_color_rgb(2, 255, 255)._data
{'code': 1000}
>>> ctr1.get_led_color()._data
{'hue': 180, 'saturation': 253, 'value': 255, 'red': 2, 'green': 255, 'blue': 255, 'code': 1000}
>>> ctr1.set_led_color_rgb(4, 255, 255)._data
{'code': 1000}
>>> ctr1.get_led_color()._data
{'hue': 180, 'saturation': 251, 'value': 255, 'red': 4, 'green': 255, 'blue': 255, 'code': 1000}
>>> 
>>> ctr2.firmware_version()._data
{'version': '2.8.3', 'code': 1000}
>>> ctr2.set_led_color_rgb(0, 255, 255)._data
{'code': 1000}
>>> ctr2.get_led_color()._data
{'hue': 180, 'saturation': 254, 'value': 255, 'red': 1, 'green': 255, 'blue': 255, 'code': 1000}
>>> ctr2.set_led_color_rgb(2, 255, 255)._data
{'code': 1000}
>>> ctr2.get_led_color()._data
{'hue': 180, 'saturation': 252, 'value': 255, 'red': 3, 'green': 255, 'blue': 255, 'code': 1000}
>>> ctr2.set_led_color_rgb(4, 255, 255)._data
{'code': 1000}
>>> ctr2.get_led_color()._data
{'hue': 180, 'saturation': 250, 'value': 255, 'red': 5, 'green': 255, 'blue': 255, 'code': 1000}
```
The previous firmware is consistent and gives the expected results. The new firmware *adds one* to the smallest rgb coordinate ("red" in this case) compared to what I asked for, and correspondingly *decreases* the expected saturation by one. This seems to happen when I set red and green to 255. When I use a lower number, say 250, it behaves better.

What I have done now: I found two settings, one of rgb and one of hsv, which actually behaves the same on both devices and gives the intuitively correct result on both. I have modified the unit test to use those values, and have run it on both my devices, without cassettes (but without isrecording), and it passes both here. I will check it in.

(How to carefully design a unit test not to trip a bug... Well, it is not our task to detect inconsistencies in the firmware, only in our code.)

Anders-Holst committed:
Adjust color unit test to work also on newest firmware

Anders-Holst committed:
Sorry, wrong ip-address...

scrool commented:
Unfortunately my testing Twinkly device is dead now. I would prefer not to screw up config on other devices or even brick them with firmware update before Christmas. If I won't get that broken one back through warranty claim this week I'm afraid this will have to wait until January or so. :frowning_face:

Anders-Holst commented:
@scrool, that's sad to hear. That it is dead I mean. The color mode is not at all urgent - I have the xled_plus to play with (most of my free time goes creating new daily effects... I should have anticipated this of course), and people can access the high level functionality they need through there, so no need to hurry here.

I was also thinking that maybe it would be good to investigate how RGBW and AWW profile devices behave regarding color mode. For RGBW there might be a white component too in the call, but how it comes into the translation to or from HSV is unclear. Whether the concept of HSV is at all meaningful for AWW is even more unclear to me.

By the way, how did you manage to kill your test device? The firmware downgrading appears to me to be a risk moment. Especially since factory reset won't reset the firmware version. If I remember there were some bugs in the current update_firmware in xled which makes it not work on at least generation II, and my proposed variant in the now withdrawn PR #80, designed to work for generation II, is never tested. But maybe it was something completely unrelated?

scrool commented:
> I was also thinking that maybe it would be good how RGBW and AWW profile devices behave regarding color mode. For RGBW there might be a wite component too, but how it comes into translation to HSV is unclear. Whether the concept of HSV is at all meaningful for AWW is unclear to me.

I can gather those data, most likely only in January.

> 
> By the way, how did you manage to kill your test device?

I have unpublished code where CLI accepts just one file, calls unpublished `firmware_update()` that accepts single file. I wanted to test one issue against old firmware version. Well, and firmware downgrade didn't suceeded - from what I can tell device never went from bootloader to boot main firmware. With Gen I I managaed to flash any firmware repeatively (<10 times) without any issues. And even when I managed to flash wrong one I was able to flash correct one on very low level (UART). I'm not sure if I did downgrade with Gen II before though. Upgrades with the same code worked just fine - that's how I gradually updated all my Gen II devices.

> The firmware downgrading appears to me to be a risk moment. Especially since factory reset won't reset the firmware version.

That sounds strange. I'd say that is part of the FW itself and independent of FW version. But I can't test that one.

> If I remember there were some bugs in the current firmware_update in xled which makes it not work on at least generation II, and my proposed variant in the now withdrawn PR #80, designed to work for generation II, is never tested. But maybe it was something completely unrelated?

I guess next time I'll reset my device first before I try to upgrade. But I'll be more careful not break them before Christmas :)

scrool commented:
I got my test device back from warranty claim. It was replaced to new one. Interestingly it has FW version 2.6.2 which I haven't seen before.

Anders-Holst commented:
That's great! 
Then I suppose you will have to play with it for a while to detect any differences in the protocol - run it through the unit tests we set up for example - before upgrading to the latest version and test color mode. Unless 2.6.2 already has color mode of course. Does it? Then we have to adjust the docs.

scrool commented:
> That's great! Then I suppose you will have to play with it for a while to detect any differences in the protocol - run it through the unit tests we set up for example - before upgrading to the latest version and test color mode. Unless 2.6.2 already has color mode of course. Does it? Then we have to adjust the docs.

No, it doesn't have `color` and you are right that I would like to ensure that it somehow works with the current firmware. AFAIK that's not something that were ever available seprately.

Anders-Holst commented:
Hi again @scrool,

I have been silent for a while, when I have taken some pause from xled - I got kind of exhausted in december from creating and publishing a new effect every day, but it was great fun too. I noticed that you meanwhile made some good changes and cleaned up a bit. Great!

Regarding color mode, the last thing we said was that we needed to test how it works on RGBW and AWW devices, to see if we need to both update the docs and complement the code in this PR. You said you could do that after Christmas. Have you had a chance to test it? That is, see what is returned by GET on the led/color endpoint on those devices. Once we have this information, I have a hypothesis we can test for how it probably works, which will enable us to adjust the code.

Ok, my hypotheis is like this: Hue-Saturation-Value makes no sense on an AWW device, so either it does not support color mode at all, or it only supports the three keywords "amber", "warmwhite" and "coldwhite" or similar with the same meaning. On RGBW, Hue-Saturation-Value makes sense but white is strictly not needed, so either it works exactly the same as RGB, or it has an extra keyword "white" which if set will internally be translated into Hue-Saturation-Value and thereafter ignored. But it will be interesting to see whether I am correct.

scrool commented:
> Regarding color mode, the last thing we said was that we needed to test how it works on RGBW and AWW devices, to see if we need to both update the docs and complement the code in this PR. You said you could do that after Christmas. Have you had a chance to test it? That is, see what is returned by GET on the led/color endpoint on those devices. Once we have this information, I have a hypothesis we can test for how it probably works, which will enable us to adjust the code.

I guess there must have been some misunderstanding. I don't have any AWW device and because of that neither `xled` nor `xled-docs` supports features that are AWW-specific. I did, however, get to test this against RGBW running on the latest firmware to date (2.8.4) and with your latest updates it indeed works.

Since I have made some changes to the code base I have rebased this PR, fixed merge conflict and updated it a little bit which ended up to be #133. And I just merged it.

Anders-Holst commented:
Aha, yes it was mainly RGBW we discussed, so maybe I extrapolated it a little.

Great that it worked and that you pushed it!

  Anders
________________________________
From: Pavol Babinčák‏ ***@***.***>
Sent: Saturday, March 26, 2022 4:20 PM
To: scrool/xled ***@***.***>
Cc: Anders Holst ***@***.***>; Author ***@***.***>
Subject: Re: [scrool/xled] Color mode (PR #113)


Regarding color mode, the last thing we said was that we needed to test how it works on RGBW and AWW devices, to see if we need to both update the docs and complement the code in this PR. You said you could do that after Christmas. Have you had a chance to test it? That is, see what is returned by GET on the led/color endpoint on those devices. Once we have this information, I have a hypothesis we can test for how it probably works, which will enable us to adjust the code.

I guess there must have been some misunderstanding. I don't have any AWW device and because of that neither xled nor xled-docs supports features that are AWW-specific. I did, however, get to test this against RGBW running on the latest firmware to date (2.8.4) and with your latest updates it indeed works.

Since I have made some changes to the code base I have rebased this PR, fixed merge conflict and updated it a little bit which ended up to be #133<https://eur05.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fscrool%2Fxled%2Fpull%2F133&data=04%7C01%7Canders.holst%40ri.se%7C9c89cbe552b942a4124708da0f3c2a76%7C5a9809cf0bcb413a838a09ecc40cc9e8%7C0%7C0%7C637839048355611576%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=UD0ECrBWABY1E3PZu16%2BD4umHZDJX0xvXZnRmADAdRA%3D&reserved=0>. And I just merged it.

—
Reply to this email directly, view it on GitHub<https://eur05.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fscrool%2Fxled%2Fpull%2F113%23issuecomment-1079714644&data=04%7C01%7Canders.holst%40ri.se%7C9c89cbe552b942a4124708da0f3c2a76%7C5a9809cf0bcb413a838a09ecc40cc9e8%7C0%7C0%7C637839048355611576%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=XVxQIrFpqKO6OMI5BFG2D3DXx1%2FwAZpXMh0Jk6ZkaXE%3D&reserved=0>, or unsubscribe<https://eur05.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAPRE5C7XWRUWP7NLBJHPOWLVB4TL3ANCNFSM5JLXER5A&data=04%7C01%7Canders.holst%40ri.se%7C9c89cbe552b942a4124708da0f3c2a76%7C5a9809cf0bcb413a838a09ecc40cc9e8%7C0%7C0%7C637839048355611576%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=9F2B906BI6%2BMwwr5aJ%2F7PHTKBZfmNN1muBQYpQRxcJI%3D&reserved=0>.
You are receiving this because you authored the thread.Message ID: ***@***.***>
